tokio-util/src/taskset.rs

```diff
@@ -0,0 +1,347 @@
+
+use futures_util::FutureExt;
+use std::future::Future;
+use std::task::{Context, Poll};
+use tokio::runtime::Handle;
+use tokio::task::{JoinError, JoinHandle};
+
+/// A [`TaskSet`] is a mechanism for regulating the lifetimes of a set of asynchronous tasks.
+///
+/// # Task Lifetimes
+/// Tasks spawned on a [`TaskSet`] live at most slightly longer than the set that spawns them.
+/// The reason for this is that aborting a task is not an instantaneous operation.
+///
+/// Tasks could be running in another worker thread at the time they are cancelled, preventing
+/// them from being canceled immediately. As a result, when you abort those tasks using the join
+/// handles, Tokio marks those futures for cancellation. There are, however ways to await the
+/// cancellation of tasks, but they only work in an async context.
+///
+/// If you need to cancel a set and wait for the cancellation to complete, use [`Self::shutdown`].
+/// That function won't complete until *all* tasks have exited.
+#[derive(Debug)]
+pub struct TaskSet<T> {
+    tasks: Vec<JoinHandle<T>>,
+    handle: Handle,
+}
+
+impl<T> TaskSet<T> {
+    /// Constructs a new TaskSet.
+    ///
+    /// # Panic
+    /// Panics if invoked outside the context of a tokio runtime.
+    pub fn new() -> Self {
+        Self::new_with_handle(Handle::current())
+    }
+
+    /// Constructs a new TaskSet which will spawn tasks using the supplied handle.
+    pub fn new_with_handle(handle: Handle) -> Self {
+        Self {
+            tasks: Vec::new(),
+            handle,
+        }
+    }
+
+    /// Returns the amount of unjoined tasks in the set.
+    pub fn len(&self) -> usize {
+        self.tasks.len()
+    }
+
+    /// Returns true if there are no tasks left in the set.
+    pub fn is_empty(&self) -> bool {
+        self.tasks.is_empty()
+    }
+
+    /// Consume the set, waiting for all tasks on it to complete.
+    ///
+    /// # Output
+    /// The output of this function orders the results of tasks in the same order with which they
+    /// were spawned.
+    ///
+    /// # Cancellation Safety
+    /// When canceled, any tasks which have yet to complete will become detached, and the results
+    /// of completed tasks will be discarded.
+    ///
+    /// # Examples
+    /// ```
+    /// # use tokio_util::taskset::TaskSet;
+    /// # tokio_test::block_on(async {
+    /// let mut set = TaskSet::new();
+    ///
+    /// const NUMS: [u8; 5] = [0, 1, 2, 3, 4];
+    ///
+    /// for x in NUMS.iter() {
+    ///     set.spawn(async move { *x });
+    /// }
+    ///
+    /// let joined: Vec<_> = set.join_all().await.into_iter().map(|x| x.unwrap()).collect();
+    ///
+    /// assert_eq!(NUMS, joined.as_slice());
+    /// # });
+    /// ```
+    pub async fn join_all(mut self) -> Vec<Result<T, JoinError>> {
+        let mut output = Vec::with_capacity(self.tasks.len());
+
+        for task in self.tasks.iter_mut() {
+            output.push(task.await)
+        }
+
+        output
```

Comment:
Seems like it should be ok to add an `self.tasks.clear()` after the loop so the destructor doesn't have to abort everything again.

Reply:
Done.